feat(#1281): Add a doc for secure user management at scale (#1283)
medic/cht-docs: content/en/apps/guides/security/securely-onboarding-users-at-scale.md

---
title: "Securely onboarding users at scale"
linkTitle: "Securely Onboarding Users"
weight:
description: >
    How to securely create users, handle password changes and password breaches
relatedContent: >
    apps/guides/training/onboarding
    apps/guides/database/couchdb-authentication
    apps/tutorials/contact-and-users-1
    apps/guides/data/users-bulk-load
    apps/concepts/users
    apps/concepts/access/#magic-links-for-logging-in-token-login

---

{{% pageinfo %}}

> The more secure a system is, the harder it is to use. Following this document will achieve the highest level of security. That said, implementers need to know when ease of use is more important than a more secure system. By reading this document you should be able to know when to make the "more secure" vs "easier to use" trade off. 

Author: Excellent change!  Same result but with a much more friendly approach - thanks!

No system is perfectly secure - be prepared to remediate a security breach!
{{% /pageinfo %}}

> When a CHT deployment will support 100s or even 1000s of users, secure credential management becomes critical. Assuming there's a 10:1 ratio of 4-person families per CHW, this means there's potentially over 40,000 patients in a large CHT instance. Their data is sensitive and should never be viewed, or worse edited, by any unauthorized parties. By following best practices, being aware of and prepared for a worst case scenario of a password breach, patient data can be kept safe and CHWs can be kept online and able to deliver care.

Reviewer: The patient calculation was based on a few assumptions and I don't think it's necessary.

```suggestion
When a CHT deployment will support hundreds of users or more, secure credential management becomes critical. Their data is sensitive and should never be viewed, or worse edited, by any unauthorized parties. By following best practices, being aware of and prepared for a worst case scenario of a password breach, patient data can be kept safe and CHWs can be kept online and able to deliver care.
```

## Before starting

### Secure devices

> It is assumed that the CHWs' devices are secure: they all employ disk encryption and require a password or PIN to unlock and use. Please see our [Securing Android Devices document]({{% ref "apps/guides/security/securing-android" %}}) for more information. As well, an [MDM]({{% ref "apps/guides/android/publishing#mobile-device-management" %}}) may be used to enforce disk encryption and device unlock protocols.

Reviewer: Let's not assume that...

```suggestion
Firstly, ensure that the CHWs' devices are secure: they all employ disk encryption and require a password or PIN to unlock and use. Please see our [Securing Android Devices document]({{% ref "apps/guides/security/securing-android" %}}) for more information. As well, an [MDM]({{% ref "apps/guides/android/publishing#mobile-device-management" %}}) may be used to enforce disk encryption and device unlock protocols.
```

### Secure administrative users

CHT administrators have the ability to create and delete users and push new configurations to the CHT should take extra precautions in managing their password. They should use a [strong passphrase](https://en.wikipedia.org/wiki/Passphrase) (instead of a password) that is unique to their CHT login. They should use a [password manager](https://en.wikipedia.org/wiki/Password_manager) to store this password. 

> By not storing their password in [plaintext](https://en.wikipedia.org/wiki/Plaintext) and by using a strong passphrase, unauthorized users are more likely to be kept out of their CHT user. 

Reviewer: This is hard to read - I think there's a double negative in there that's making it complex.

```suggestion
By following these steps, unauthorized people are less likely to be able to access administrator accounts. 
```

## Spreadsheet use

Most deployments manage users in a spreadsheet shared either in Google Docs or other cloud service. It is convenient to have a canonical shared location to access the data. This is an acceptable, but not ideal, solution as it ensures user account changes are instantly shared with all CHT admins, but can keep a number of key security requirements. For an ideal solution, [see "Ideal" below](#ideal-1-only-magic-links).

Of paramount importance:

* **Never** enable anonymous access to the URL for the spreadsheet. You **must always require authentication** to access user credential lists
* Audit regularly who has access to the shared spreadsheet. Consider setting up an alert every time a new person is granted access.

> * Reduce the total number of users that need access by breaking up user lists by logical group. For example each sub-county could have its own user list. Only admins working for with sub-county would have access to the user list they work with.

Reviewer: ```suggestion
* Reduce the total number of users that need access by breaking up user lists by logical group. For example each sub-county could have its own user spreadsheet. That way the user credentials are only shared with the local administrators who need it.
```

* Refrain from printing spreadsheets of users and passwords. They can be lost, stolen or easily photographed when shown in public.

> * If you never need to download a plaintext CSV with username and password, ensure the computer also has disk encryption enabled and requires a password to unlock.

Reviewer: ```suggestion
* If you ever need to download a plaintext CSV with username and password, ensure the computer also has disk encryption enabled and requires a password to unlock.
```

## Transmitting credentials

TK - only put in spreadsheet and share direct access. Do not send over Slack, Whatsapp, email or other medium. Use magic link instead.

TK - when not using magic link, what is the realistic way to send credentials to a single user? 

TK - when not using magic link, what is the realistic way to bulk provision 100s of devices without a printed list you can cross off? 

TK - notes about downloading creds from [user management tool](https://github.com/medic/cht-user-management/) per [slack thread](https://medic.slack.com/archives/CHYAGKHN2/p1706744894943699?thread_ts=1706728984.849139&cid=CHYAGKHN2)? 

## Example scenarios

### Ideal practice 1: Only Magic Links

Create a spreadsheet with full name, telephone number etc. Included is a username but NOT a password. When users are created in bulk via the command line or bulk user upload, have magic links sent to the user via an SMS gateway. This avoids the problem of passwords being stored in clear text in the spreadsheet or on a printed version.

### Ideal practice 2: Unknown passwords, reset during provision

> An alternate and also secure approach, is to bulk create the users as described above, not use magic links, and use random passwords that you do not save. At a later time, manually change each user's password as you provision their mobile device. This also prevents a list of passwords from being stored in clear text.

Reviewer: ```suggestion
An alternate and also secure approach, is to bulk create the users as described above, not use magic links, and use random passwords that you do not save. Train the users on changing their password after they've logged in for the first time. This ensures if the initial password has leaked it's now no longer useful.
```

Author: ah interesting!   your edit involves the user, my original does not.  My thinking, and it may not scale to the field when provisioning 100s of devices, is to first bulk create all users then, as your provisioning each device, change the password one at a time. This is done entirely by the admins, and not known by the user.

Thoughts on which is more viable?

Reviewer: When discussing with a large partner recently, we discovered that changing the password was already something that was included in training, although they just showed the screen and didn't actually submit the change. So it would fit nicely into their training to fill it in while they're there. For any partner who doesn't already include it it would be good to add because once they've done it if they ever need to change their password due to a breach they'll remember how to do it from last time.

Author: Gotcha - makes sense!  Let's go with your version.  thanks!

* Do not keep downloaded CSV plaintext copies of credentials. Instead, delete and re-download them from the authorized, authenticated cloud server as needed

## Transmitting credentials

When it comes time get a username and password on to a device or to a remote user, be sure to use mediums that are secure. One of the main concerns is credentials being found long after they were sent.

* A best practice is for the sender to add a credential to a shared password manager.  The person receiving the credentials can then securely open the password manager. 
* If no password manager is available, consider sending the password via [One Time Secret](https://onetimesecret.com/)
* To send credentials in to many CHWs, consider using [token login](https://docs.communityhealthtoolkit.org/apps/concepts/access/#magic-links-for-logging-in-token-login).
* For sending large lists of credentials, as mentioned above, using a cloud provider like Google Sheets, is a good way to have an audit trail and still provide easy, remote access.

## Example scenarios

### Ideal practice 1: Only Magic Links

> Create a spreadsheet with all your users' data. Included is a username but NOT a password. When users are created in bulk via the command line or bulk user upload, have a [token login](https://docs.communityhealthtoolkit.org/apps/concepts/access/#magic-links-for-logging-in-token-login) sent to the user via an SMS gateway. This avoids the problem of passwords being stored in clear text in the spreadsheet or on a printed version. The token login links can only be used once and are only valid for 24 hours be default.

Author: This is the only place we mention this, so I think it's reasonable to exclude the "by default" (nee `be default`) unless we want to document the procedure to extend token login links. 

But, FYI,  I tested it and it is not super easy, but you can edit the entry in CouchDB to expire after any time you'd like.   Here's a token that looks like this:

```
  "token_login": {
    "active": false,
    "token": "FomT:5fU-oL1.jpFanMejTF.zfUUl5$VH-o33MCt6Dod!wXrtKTy=QNQCw0pjfN1",
    "expiration_date": 1658514078000,
    "login_date": 1658514197091
  }
```

And  you can it in couch to be valid and expire at the end of the month with:

```
  "token_login": {
    "active": true,
    "token": "FomT:5fU-oL1.jpFanMejTF.zfUUl5$VH-o33MCt6Dod!wXrtKTy=QNQCw0pjfN1",
    "expiration_date": 1690050078000
  }
```

    -d '{"token_login":true}'
```

### Audit who has changed their password

If you have instructed all users to manually change their password, CHT administrators can search their server logs to see which users have made the update.  First find a list of all running services that have the word `haproxy` in it:

```shell
docker ps --filter "name=haproxy" --format '{{.Names}}'
```

Then, assuming your service is called `cht_haproxy_1`, run this command to find all password changes in the past 24 hours:

```shell
docker logs cht_haproxy_1 --since 24h 2>&1  | grep password_scheme
```

Here is an example event showing user `mary` having changed their password on Feb 9 18:47:

`
Feb  9 18:47:08 haproxy[12]: 172.22.0.2,couchdb,201,15,0,0,PUT,/_users/org.couchdb.user%3Amary,api,medic,'{"_id":"org.couchdb.user:mary","_rev":"3-d3dd14f3026942245ce3881adfcd513e","name":"mary","type":"user","roles":["chw"],"facility_id":"14046008-f796-4418-9ff2-12c2ef77b478","password_scheme":"***","iterations":10,"derived_key":"f846a7a20209592c613c09ea4405170735c7a557","salt":"bbb089e5fc459c2ca1088c8316e0cc99","password":"***"}',411,15,86,'node-fetch/1.0 (+https://github.com/bitinn/node-fetch)'

> `

Author: so I actually intentionally did not go for   <code>```</code> opting for <code>`</code> . This is because <code>`</code> gives us the mono-space font with different color background but wraps. the line is so comically long w/o the wrap that I figured it was worth the non-traditional use of <code>`</code>.  Further the string I'm interested in is about 1/2 down the string, make it it hard to sideways scroll to find.
